From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mariell Hoversholm <proximyst@proximyst.com>
Date: Sat, 9 Jan 2021 16:06:40 +0100
Subject: [PATCH] Purpur: Farmland trampling changes

Original license: MIT
Original project: https://github.com/PurpurMC/Purpur

This lets us choose if farmland trampling is fully disabled or only
players may trample farmland.

This lets us choose if entities can stop trampling if they fall a
distance equal to their feather falling level, plus the extra block
necessary to trample in the first place. Feather Falling 1 requires
you to fall over 3+ blocks to trample. FF 2 requires 4+, etc.

diff --git a/src/main/java/net/minecraft/world/level/block/FarmBlock.java b/src/main/java/net/minecraft/world/level/block/FarmBlock.java
index 1887e14a99b70841b95c70af87a1ee9b0e2b3802..17ea6dfddff2d40ca7343b3f5eafa12d7d75e8c8 100644
--- a/src/main/java/net/minecraft/world/level/block/FarmBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/FarmBlock.java
@@ -113,6 +113,16 @@ public class FarmBlock extends Block {
                 return;
             }
 
+            // Purpur start
+            if (LeafConfig.farmlandTramplingDisabled) return;
+            if (LeafConfig.farmlandTramplingOnlyPlayers && !(entity instanceof Player)) return;
+            if (LeafConfig.farmlandTramplingFeatherFalling) {
+                Iterator<net.minecraft.world.item.ItemStack> armor = entity.getArmorSlots().iterator();
+                if (armor.hasNext() && net.minecraft.world.item.enchantment.EnchantmentHelper.getItemEnchantmentLevel(net.minecraft.world.item.enchantment.Enchantments.FALL_PROTECTION, armor.next()) >= (int) entity.fallDistance) {
+                    return;
+                }
+            }
+            // Purpur end
             if (CraftEventFactory.callEntityChangeBlockEvent(entity, pos, Blocks.DIRT.defaultBlockState()).isCancelled()) {
                 return;
             }
diff --git a/src/main/java/org/dreeam/leaf/LeafConfig.java b/src/main/java/org/dreeam/leaf/LeafConfig.java
index 9f3ef32caad977987f934a25053993bcb1a5b2a7..589254410a7afc66a9ec0f374ba48a0574ca11fb 100644
--- a/src/main/java/org/dreeam/leaf/LeafConfig.java
+++ b/src/main/java/org/dreeam/leaf/LeafConfig.java
@@ -360,6 +360,9 @@ public class LeafConfig {
     public static boolean farmlandBypassMobGriefing = false;
     public static boolean powderSnowBypassMobGriefing = false;
     public static boolean turtleEggsBypassMobGriefing = false;
+    public static boolean farmlandTramplingDisabled = false;
+    public static boolean farmlandTramplingOnlyPlayers = false;
+    public static boolean farmlandTramplingFeatherFalling = false;
     private static void blockSettings() {
         spongeAbsorptionArea = getInt("blocks.sponge.absorption.area", spongeAbsorptionArea);
         spongeAbsorptionRadius = getInt("blocks.sponge.absorption.radius", spongeAbsorptionRadius);
@@ -409,6 +412,9 @@ public class LeafConfig {
         farmlandBypassMobGriefing = getBoolean("blocks.farmland.bypass-mob-griefing", farmlandBypassMobGriefing);
         powderSnowBypassMobGriefing = getBoolean("blocks.powder_snow.bypass-mob-griefing", powderSnowBypassMobGriefing);
         turtleEggsBypassMobGriefing = getBoolean("blocks.turtle_egg.bypass-mob-griefing", turtleEggsBypassMobGriefing);
+        farmlandTramplingDisabled = getBoolean("blocks.farmland.disable-trampling", farmlandTramplingDisabled);
+        farmlandTramplingOnlyPlayers = getBoolean("blocks.farmland.only-players-trample", farmlandTramplingOnlyPlayers);
+        farmlandTramplingFeatherFalling = getBoolean("blocks.farmland.feather-fall-distance-affects-trampling", farmlandTramplingFeatherFalling);
     }
 
     public static boolean kickForOutOfOrderChat = true;
